### PR TITLE
Compacting: integrate into Flows system

### PR DESCRIPTION
**Pull request by rmn-boiko**
## Description

Closes: [Compacting: integrate into Flows system](https://github.com/kamu-data/kamu-cli/issues/583)

## Checklist before requesting a review

- [x] [CHANGELOG.md](./CHANGELOG.md) updated
- [x] API changes are backwards-compatible
- [x] Workspace layout changes include a migration
- [x] Documentation update PR: <link or N/A>
- [x] Dataset pipelines update scheduled if needed
- [x] Unit-tests added

### CODE REVIEW

File: src/domain/flow-system/src/entities/shared/flow_key.rs [hunk 1]
```diff
@@ -34,13 +35,23 @@ impl FlowKey {
 pub struct FlowKeyDataset {
     pub dataset_id: DatasetID,
     pub flow_type: DatasetFlowType,
+    pub options: CompactOptions,
```
Comment (zaychenko-sergei) [inline src/domain/flow-system/src/entities/shared/flow_key.rs:38]: Logically this is unrelated to keys at all.
I need to think where to put this data, it's rather a part of flow configuration.
For now, pass the defaults to the service.

File: src/adapter/graphql/src/mutations/flows_mut/dataset_flow_runs_mut.rs [hunk 2]
```diff
@@ -42,6 +42,8 @@ impl DatasetFlowRunsMut {
         &self,
         ctx: &Context<'_>,
         dataset_flow_type: DatasetFlowType,
+        max_slice_size: Option<u64>,
+        max_slice_records: Option<u64>,
```
[the remark this answers is not in the document]
Comment (rmn-boiko) [reply src/adapter/graphql/src/mutations/flows_mut/dataset_flow_runs_mut.rs]: Removed

File: src/domain/flow-system/src/entities/shared/flow_key.rs (hunk 1 above)
Comment (rmn-boiko) [reply src/domain/flow-system/src/entities/shared/flow_key.rs]: Removed as well

File: src/infra/flow-system-inmem/tests/tests/test_flow_service_inmem.rs [hunk 3]
```diff
@@ -414,6 +426,293 @@ async fn test_manual_trigger() {
...
+        _ = async {
+                // Task 0: "foo" start running at 10ms, finish at 20ms
+                let task0_driver = harness.task_driver(TaskDriverArgs {
+                    task_id: TaskID::new(0),
+                    dataset_id: Some(foo_id.clone()),
+                    run_since_start: Duration::try_milliseconds(10).unwrap(),
+                    finish_in_with: Some((Duration::try_milliseconds(10).unwrap(), TaskOutcome::Success(TaskResult::Empty))),
+                    max_slice_size: None,
+                    max_slice_records: None,
+                });
```
Comment (zaychenko-sergei) [inline src/infra/flow-system-inmem/tests/tests/test_flow_service_inmem.rs:477]: No tasks would auto-start for compaction flow - you are not auto-scheduling anything for this type.

File: src/infra/flow-system-inmem/tests/tests/test_flow_service_inmem.rs [hunk 4]
```diff
@@ -414,6 +426,293 @@ async fn test_manual_trigger() {
...
+                // Task 0: "foo" start running at 10ms, finish at 20ms
+                let task0_driver = harness.task_driver(TaskDriverArgs {
+                    task_id: TaskID::new(0),
+                    dataset_id: Some(foo_id.clone()),
+                    run_since_start: Duration::try_milliseconds(10).unwrap(),
+                    finish_in_with: Some((Duration::try_milliseconds(10).unwrap(), TaskOutcome::Success(TaskResult::Empty))),
+                    max_slice_size: Some(max_slice_size),
+                    max_slice_records: Some(max_slice_records),
+                });
+                let task0_handle = task0_driver.run();
```
Comment (zaychenko-sergei) [inline src/infra/flow-system-inmem/tests/tests/test_flow_service_inmem.rs:667]: Why would task start before the trigger? If trigger is at time 40ms, the task should go after, maybe at 50ms.

File: src/infra/flow-system-inmem/tests/tests/utils/task_driver.rs [hunk 5]
```diff
@@ -30,6 +30,8 @@ pub(crate) struct TaskDriverArgs {
     pub(crate) dataset_id: Option<DatasetID>,
     pub(crate) run_since_start: Duration,
     pub(crate) finish_in_with: Option<(Duration, TaskOutcome)>,
+    pub(crate) max_slice_size: Option<u64>,
+    pub(crate) max_slice_records: Option<u64>,
```
Comment (zaychenko-sergei) [inline src/infra/flow-system-inmem/tests/tests/utils/task_driver.rs:34]: Why is this a part of task driver? We should define parameters at flow configuration level only.

**Review by zaychenko-sergei — CHANGES_REQUESTED** [review]
I don't see where you have addressed this part of requirements:

image

File: src/adapter/graphql/src/scalars/flow_configuration.rs [hunk 6]
```diff
@@ -64,6 +78,21 @@ impl From<BatchingRule> for FlowConfigurationBatching {
     }
 }
 
+#[derive(SimpleObject, Clone, PartialEq, Eq)]
+pub struct FlowConfigurationCompacting {
+    pub max_slice_size: u64,
+    pub max_slice_records: u64,
+}
+
+impl From<CompactionRule> for FlowConfigurationCompacting {
```
[the remark this answers is not in the document]
Comment (zaychenko-sergei) [reply src/adapter/graphql/src/scalars/flow_configuration.rs:87]: I mean everywhere in the code, not just here

File: src/adapter/graphql/src/mutations/flows_mut/dataset_flow_configs_mut.rs [hunk 7]
```diff
@@ -73,6 +75,14 @@ impl DatasetFlowConfigsMut {
...
             }
         };
+        if !ensure_set_config_flow_supported(
+            dataset_flow_type,
+            &FlowConfigurationRule::Schedule(configuration_rule.clone()),
+        ) {
+            return Ok(SetFlowConfigResult::TypeIsNotSupported(
+                FlowTypeIsNotSupported,
+            ));
+        }
```
[the remark this answers is not in the document]
Comment (rmn-boiko) [reply src/adapter/graphql/src/mutations/flows_mut/dataset_flow_configs_mut.rs:85]: In this case we will need to use type of our enum, but I agree this check should be above.

File: src/adapter/graphql/src/scalars/flow_configuration.rs (hunk 6 above)
[the remark this answers is not in the document]
Comment (rmn-boiko) [reply src/adapter/graphql/src/scalars/flow_configuration.rs:87]: Done

File: src/domain/core/src/services/compact_service.rs [hunk 8]
```diff
@@ -9,18 +9,21 @@
...
+pub const DEFAULT_MAX_SLICE_RECORDS: u64 = 10000;
+
 #[async_trait::async_trait]
 pub trait CompactService: Send + Sync {
     async fn compact_dataset(
         &self,
         dataset_handle: &DatasetHandle,
-        max_slice_size: u64,
-        max_slice_records: u64,
+        options: &CompactOptions,
```
[the remark this answers is not in the document]
Comment (rmn-boiko) [reply src/domain/core/src/services/compact_service.rs:26]: agree, removed pointer

File: src/infra/core/src/compact_service_impl.rs [hunk 9]
```diff
@@ -17,12 +17,15 @@ use datafusion::prelude::*;
...
     CompactError,
+    CompactOptions,
     CompactResult,
     CompactionListener,
     CompactionMultiListener,
     CompactionPhase,
     InvalidDatasetKindError,
     NullCompactionListener,
+    DEFAULT_MAX_SLICE_RECORDS,
+    DEFAULT_MAX_SLICE_SIZE,
```
[the remark this answers is not in the document]
Comment (rmn-boiko) [reply src/infra/core/src/compact_service_impl.rs:28]: Will add default implementation with these value, but still think we need to keep this fallback in case of None from any reason

File: src/infra/flow-system-inmem/src/services/flow/flow_service_inmem.rs [hunk 10]
```diff
@@ -662,7 +663,26 @@ impl FlowServiceInMemory {
...
+                );
+            if let Some(opts) = maybe_compaction_rule {
+                logical_plan_opts = LogicalPlanOptions {
+                    max_slice_size: Some(opts.max_slice_size()),
+                    max_slice_records: Some(opts.max_slice_records()),
+                };
+            };
+        }
+
+        let logical_plan = flow.make_task_logical_plan(&logical_plan_opts);
```
[the remark this answers is not in the document]
Comment (rmn-boiko) [reply src/infra/flow-system-inmem/src/services/flow/flow_service_inmem.rs:685]: Done

File: src/infra/flow-system-inmem/tests/tests/utils/task_driver.rs (hunk 5 above)
Comment (rmn-boiko) [reply src/infra/flow-system-inmem/tests/tests/utils/task_driver.rs:34]: It's here just to assert we are using the right parameters. Not sure I can do it in another way

File: src/infra/flow-system-inmem/tests/tests/test_flow_service_inmem.rs (hunk 4 above)
Comment (rmn-boiko) [reply src/infra/flow-system-inmem/tests/tests/test_flow_service_inmem.rs:667]: Fixed

File: src/infra/flow-system-inmem/tests/tests/utils/task_driver.rs (hunk 5 above)
Comment (rmn-boiko) [reply src/infra/flow-system-inmem/tests/tests/utils/task_driver.rs:34]: Changed to assert logical plan

File: src/adapter/graphql/src/scalars/flow_scalars.rs [hunk 11]
```diff
@@ -100,22 +100,22 @@ pub enum FlowStatus {
...
+
+// impl From<&kamu_flow_system::FlowOutcome> for FlowOutcome {
+//     fn from(value: &kamu_flow_system::FlowOutcome) -> Self {
+//         match value {
+//             kamu_flow_system::FlowOutcome::Success(_) => Self::Success,
+//             kamu_flow_system::FlowOutcome::Failed(err) => Self::Failed(err),
+//             kamu_flow_system::FlowOutcome::Aborted => Self::Aborted,
+//         }
+//     }
+// }
```
Comment (zaychenko-sergei) [inline src/adapter/graphql/src/scalars/flow_scalars.rs:118]: Clean this

File: src/adapter/graphql/src/scalars/flow_scalars.rs [hunk 12]
```diff
@@ -136,3 +136,11 @@ pub enum SystemFlowType {
 }
 
 /////////////////////////////////////////////////////////////////////////////////////////
+
+// #[derive(Debug, Clone, PartialEq, Eq)]
+// #[graphql(remote = "kamu_flow_system::FlowFailedError")]
+// pub struct FlowFailedError {
+//     pub reason: String,
+// }
```
Comment (zaychenko-sergei) [inline src/adapter/graphql/src/scalars/flow_scalars.rs:144]: and this

File: src/domain/flow-system/src/entities/flow/flow_outcome.rs [hunk 13]
```diff
@@ -38,13 +40,38 @@ impl FlowOutcome {
...
-            FlowResult::DatasetUpdate(_) => false,
+            FlowResult::DatasetUpdate(_) | FlowResult::DatasetCompact(_) => false,
+        }
+    }
+}
+
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub enum FlowError {
+    Failed,
+    RootDatasetWasCompacted(FlowRootDatasetWasCompactedError),
```
Comment (zaychenko-sergei) [inline src/domain/flow-system/src/entities/flow/flow_outcome.rs:58]: Let's remove "was" from the name

File: src/domain/task-system/domain/src/entities/logical_plan.rs [hunk 14]
```diff
@@ -23,6 +23,8 @@ pub enum LogicalPlan {
     UpdateDataset(UpdateDataset),
     /// A task that can be used for testing the scheduling system
     Probe(Probe),
+    /// Perform dataset hard or soft compacting
```
Comment (zaychenko-sergei) [inline src/domain/task-system/domain/src/entities/logical_plan.rs:26]: The comment mentions "soft", which is a lie

File: src/infra/flow-system-inmem/tests/tests/utils/task_driver.rs [hunk 15]
```diff
@@ -86,17 +87,19 @@ impl TaskDriver {
...
             .expect("Task does not exist yet");
 
+        assert_eq!(self.args.expected_logical_plan, task.logical_plan);
         match &task.logical_plan {
             LogicalPlan::UpdateDataset(ud) => {
                 assert!(self.args.dataset_id.is_some());
                 assert_eq!(&ud.dataset_id, self.args.dataset_id.as_ref().unwrap());
             }
             LogicalPlan::Probe(_) => assert!(self.args.dataset_id.is_none()),
+            LogicalPlan::HardCompactDataset(_) => (),
```
Comment (zaychenko-sergei) [inline src/infra/flow-system-inmem/tests/tests/utils/task_driver.rs:102]: Don't you have a dataset id  there to assert?

File: src/infra/flow-system-inmem/tests/tests/test_flow_service_inmem.rs [hunk 16]
```diff
@@ -414,6 +432,237 @@ async fn test_manual_trigger() {
...
+        res = harness.flow_service.run(start_time) => res.int_err(),
+
+        // Run simulation script and task drivers
+        _ = async {
+                  // Task 0: "foo" start running at 10ms, finish at 20ms
+                  let task0_driver = harness.task_driver(TaskDriverArgs {
+                    task_id: TaskID::new(0),
+                    dataset_id: Some(foo_id.clone()),
+                    run_since_start: Duration::try_milliseconds(10).unwrap(),
+                    finish_in_with: Some((Duration::try_milliseconds(10).unwrap(), TaskOutcome::Success(TaskResult::Empty))),
```
Comment (zaychenko-sergei) [inline src/infra/flow-system-inmem/tests/tests/test_flow_service_inmem.rs:471]: Try to have a distance between trigger and a task, or task and it's handling. This timing logic is very sensitive and may randomly overlap. If trigger happens at 10ms, start the task at 20ms and finish it at 30ms.

File: src/infra/flow-system-inmem/tests/tests/test_flow_service_inmem.rs [hunk 17]
```diff
@@ -414,6 +432,237 @@ async fn test_manual_trigger() {
...
+                    expected_logical_plan: LogicalPlan::HardCompactDataset(HardCompactDataset {
+                      dataset_id: foo_id.clone(),
+                      max_slice_size: None,
+                      max_slice_records: None,
+                    }),
+                });
+                let task0_handle = task0_driver.run();
+
+                let task2_driver = harness.task_driver(TaskDriverArgs {
+                  task_id: TaskID::new(1),
```
Comment (zaychenko-sergei) [inline src/infra/flow-system-inmem/tests/tests/test_flow_service_inmem.rs:481]: The number in `taskX_driver` should match `TaskID`, if you still plan reading your test :-)

File: src/infra/flow-system-inmem/tests/tests/test_flow_service_inmem.rs [hunk 18]
```diff
@@ -414,6 +432,237 @@ async fn test_manual_trigger() {
...
+                let trigger1_driver = harness.manual_flow_trigger_driver(ManualFlowTriggerArgs {
+                    flow_key: bar_flow_key,
+                    run_since_start: Duration::try_milliseconds(40).unwrap(),
+                });
+                let trigger1_handle = trigger1_driver.run();
+
+                // Main simulation script
+                let main_handle = async {
+                    // Moment 10ms - manual foo trigger happens here:
+                    //  - flow 0 gets trigger and is rescheduled to 40ms (20ms finish + 20ms throttling <= 40ms now)
```
Comment (zaychenko-sergei) [inline src/infra/flow-system-inmem/tests/tests/test_flow_service_inmem.rs:510]: This comment is definitely copypasted. Nothing gets rescheduled for compacting flows

File: src/infra/flow-system-inmem/tests/tests/test_flow_service_inmem.rs [hunk 19]
```diff
@@ -414,6 +432,237 @@ async fn test_manual_trigger() {
...
+
+                    // Moment 40ms - manual foo trigger happens here:
+                    //  - flow 1 trigger and is rescheduled to 40ms (20ms finish + 20ms throttling <= 40ms now)
```
Comment (zaychenko-sergei) [inline src/infra/flow-system-inmem/tests/tests/test_flow_service_inmem.rs:513]: Same here

File: src/infra/flow-system-inmem/tests/tests/test_flow_service_inmem.rs [hunk 20]
```diff
@@ -414,6 +432,237 @@ async fn test_manual_trigger() {
...
+        .unwrap();
+
+    // Run scheduler concurrently with manual triggers script
+    tokio::select! {
+        // Run API service
+        res = harness.flow_service.run(start_time) => res.int_err(),
+
+        // Run simulation script and task drivers
+        _ = async {
+                // Task 0: "foo" start running at 10ms, finish at 20ms
```
Comment (zaychenko-sergei) [inline src/infra/flow-system-inmem/tests/tests/test_flow_service_inmem.rs:605]: Careful with these comments. They are only confusing, if you don't reread and update them

File: src/infra/flow-system-inmem/tests/tests/utils/task_driver.rs (hunk 15 above)
Comment (zaychenko-sergei) [reply src/infra/flow-system-inmem/tests/tests/utils/task_driver.rs:102]: Also, where do you check the configuration options now?

File: src/infra/flow-system-inmem/src/services/flow/flow_service_inmem.rs [hunk 21]
```diff
@@ -179,6 +179,7 @@ impl FlowServiceInMemory {
                         self.enqueue_auto_polling_flow_unconditionally(start_time, &flow_key)
                             .await?;
                     }
+                    FlowConfigurationRule::CompactingRule(_) => (),
```
Comment (zaychenko-sergei) [inline src/infra/flow-system-inmem/src/services/flow/flow_service_inmem.rs:182]: Make an explaining comment that nothing should be done, only manual triggers are expected

File: src/infra/flow-system-inmem/src/services/flow/flow_service_inmem.rs [hunk 22]
```diff
@@ -717,6 +723,47 @@ impl FlowServiceInMemory {
 
         Ok(())
     }
+
+    fn get_logical_plan_options(&self, flow_key: &FlowKey) -> LogicalPlanOptions {
+        let mut logical_plan_opts = LogicalPlanOptions::default();
+        if let FlowKey::Dataset(dataset_flow) = flow_key {
```
Comment (zaychenko-sergei) [inline src/infra/flow-system-inmem/src/services/flow/flow_service_inmem.rs:729]: Where do you check the flow type is compacting and not update ?

File: src/infra/flow-system-inmem/src/services/flow/flow_service_inmem.rs [hunk 23]
```diff
@@ -717,6 +723,47 @@ impl FlowServiceInMemory {
...
+                            return TaskOutcome::Failed(TaskError::RootDatasetWasCompacted(
+                                RootDatasetWasCompactedError {
+                                    dataset_id: trigger.dataset_id.clone(),
+                                },
+                            ));
+                        }
+                    }
+                }
+            }
+        }
```
Comment (zaychenko-sergei) [inline src/infra/flow-system-inmem/src/services/flow/flow_service_inmem.rs:764]: Trick looks good!

File: src/domain/task-system/services/src/task_executor_impl.rs [hunk 24]
```diff
@@ -116,6 +117,36 @@ impl TaskExecutor for TaskExecutorImpl {
...
+                            },
+                            None,
+                        )
+                        .await;
+
+                    match compact_result {
+                        Ok(result) => {
+                            TaskOutcome::Success(TaskResult::CompactDatasetResult(result.into()))
+                        }
+                        Err(_) => TaskOutcome::Failed(TaskError::Empty),
```
Comment (zaychenko-sergei) [inline src/domain/task-system/services/src/task_executor_impl.rs:147]: What kind of error is this? Does it contain any useful details for reporting the result?

File: src/domain/task-system/domain/src/entities/logical_plan.rs [hunk 25]
```diff
@@ -58,6 +61,25 @@ pub struct Probe {
 
 /////////////////////////////////////////////////////////////////////////////////////////
 
+/// A task that can be used for testing the scheduling system
```
Comment (zaychenko-sergei) [inline src/domain/task-system/domain/src/entities/logical_plan.rs:64]: Start reading comments you copy :-)

File: src/infra/flow-system-inmem/src/services/flow/flow_service_inmem.rs [hunk 26]
```diff
@@ -662,7 +667,8 @@ impl FlowServiceInMemory {
         flow: &mut Flow,
         schedule_time: DateTime<Utc>,
     ) -> Result<TaskID, InternalError> {
-        let logical_plan = flow.make_task_logical_plan();
+        let logical_plan_opts = self.get_logical_plan_options(&flow.flow_key);
+        let logical_plan = flow.make_task_logical_plan(&logical_plan_opts);
```
Comment (zaychenko-sergei) [inline src/infra/flow-system-inmem/src/services/flow/flow_service_inmem.rs:671]: Maybe this could be a single step - take a flow key, configuration, and give a logical plan.
We can move logical plan creation into the service from the flow type if that helps.

I think if you do this, you can avoid having this additional struct `LogicalPlanOptions`, which seems to be useless.

File: src/infra/flow-system-inmem/tests/tests/utils/task_driver.rs (hunk 15 above)
Comment (rmn-boiko) [reply src/infra/flow-system-inmem/tests/tests/utils/task_driver.rs:102]: I'm asserting full logical plan above. Not sure we need to assert additionally `datase_id` here

File: src/infra/flow-system-inmem/src/services/flow/flow_service_inmem.rs (hunk 22 above)
Comment (rmn-boiko) [reply src/infra/flow-system-inmem/src/services/flow/flow_service_inmem.rs:729]: Yeah, missed it, added.

**Review by zaychenko-sergei — APPROVED** [review]
